File: engine/compose/src/commonMain/kotlin/dev/getmilano/runtime/MilanoActionHandler.kt
package dev.getmilano

/**
 * An asynchronous receiver of custom actions: one funnel per view.
 * Normal return is success and the returned value, validated against the
 * action's declared result type, binds the result root inside onSuccess;
 * return null for actions declaring no result. Throwing is failure.
 * Completion-exactly-once holds by construction.
 */
fun interface MilanoActionHandler {
    suspend fun handle(action: MilanoAction): MilanoValue?
}

/** A dispatched custom action, delivered as data. */
data class MilanoAction(
    val name: String,
    val parameters: Map<String, MilanoValue>,
    val viewIdentity: String,
)

File: engine/swiftui/Sources/MilanoSDK/Runtime/MilanoActionHandler.swift
import Foundation

/// An asynchronous receiver of custom actions: one funnel per view.
/// Normal return is success and the returned value, validated against the
/// action's declared `result` type, binds the `result` root inside
/// `onSuccess`; return `nil` for actions declaring no result. Throwing is
/// failure. Completion-exactly-once holds by construction.
public protocol MilanoActionHandler: Sendable {
    func handle(_ action: MilanoAction) async throws -> MilanoValue?
}

/// A dispatched custom action, delivered as data.
public struct MilanoAction: Equatable, Sendable {
    public let name: String
    public let parameters: [String: MilanoValue]
    public let viewIdentity: String
}

/// Closure-based convenience handler.
public struct MilanoClosureActionHandler: MilanoActionHandler {
    private let closure: @Sendable (MilanoAction) async throws -> MilanoValue?

    public init(_ closure: @escaping @Sendable (MilanoAction) async throws -> MilanoValue?) {
        self.closure = closure
    }

    public func handle(_ action: MilanoAction) async throws -> MilanoValue? {
        try await closure(action)
    }
}
